### PR TITLE
DONT MERGE: Registration form backend

### PR DESCRIPTION
**Pull request by Riscky**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,8 @@
 npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
+.idea/intro-website.iml
```
Comment (Riscky) [inline .gitignore:25]: can't we just ignore the whole .idea folder?

File: src/formBackend/RegistrationBackend.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+const { check, validationResult } = require('express-validator');
+const app = express();
+const sqlite = require('sqlite3');
+let DB = new sqlite.Database('src/formBackend/ApplicationsDatabase.db');
+
+const bodyParser = require('body-parser');
+app.use(bodyParser.urlencoded({extended:true}));
+app.use(bodyParser.json());
+
+const studyOptions = ["", "IK", "IC", "IC/IK", "NA", "WI", "GT", "WIT"];
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:11]: I don't think we have anything other than IK, IC or IC/IK students
Also empty should probably not be an option

File: src/formBackend/RegistrationBackend.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+app.post('/submitRegistration',[
+    check('fvoornaam').escape(),
+    check('fvoorletters').escape(),
+    check('ftussenvoegsel').escape(),
+    check('fachternaam').escape(),
+    check('fgeboortedatum').isDate().withMessage("Your date of birth is not valid.").escape(),
+    check('fstudentnummer').custom(studentID => validStudentID((studentID))).escape(),
+    check('fmobiel').escape(),
+    check('fstudie').isIn(studyOptions).withMessage("The study you have entered does not exist, please stop hacking our site.").escape(),
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:21]: :laughing:

File: src/formBackend/RegistrationBackend.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    check( 'femail').isEmail().escape().withMessage("Your email is not valid.").escape()
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:22]: also nice checks, we should double check if they align with the checks in the frontend

File: src/formBackend/RegistrationBackend.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        if(!errors.isEmpty()){
+            res.status(422).send(errors);
+            return
+        }
+        console.log(req.body);
+
+        addStudentToDB(req.body).then((addActionResult) =>{
+
+        console.log(addActionResult);
+        if(addActionResult == 1){
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:36]: use === instead of ==

File: src/formBackend/RegistrationBackend.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            res.status(500).send("Internal server error while processing your application. Please try again later. ")
+        }
+        else if (addActionResult == 2){
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:39]: '==='

File: src/formBackend/RegistrationBackend.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+* @param {string} reqBody sanitized body from the form.
+* */
+function addStudentToDB(reqBody){
+    let studentID = reqBody.fstudentnummer;
+    return new Promise(function (resolve, reject) {
+        let sql = 'SELECT studentnummer ';
+        sql += 'FROM Applications';
+        sql += ' WHERE studentnummer = ?';
+
+        DB.getasync(sql,studentID).then(((rows) => {if(rows.length == 0){
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:102]: ===

File: src/formBackend/RegistrationBackend.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            });
+    })
+}
+
+/**
+ * Checks if a student ID is a valid student ID.
+ * @param studentID The studentID you want to check
+ */
+function validStudentID(studentID){
+    if(studentID.length != 7){
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:128]: !==

File: src/formBackend/RegistrationBackend.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        throw new Error("StudentID does not match the required length of 7 numbers.");
+    }
+    let multiplier = 7;
+    let result = 0;
+    while(multiplier > 0){
+        result +=  parseInt(studentID[7 - multiplier]) * multiplier;
+        multiplier--;
+    }
+    if(result % 11 != 0){
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:137]: !==

File: src/formBackend/RegistrationBackend.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        throw new Error("StudentID is not a valid student ID.");
+        return false;
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:139]: unreachable return statement :)

File: src/formBackend/RegistrationBackend.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    let studentID = reqBody.fstudentnummer;
+    return new Promise(function (resolve, reject) {
+        let sql = 'SELECT studentnummer ';
+        sql += 'FROM Applications';
+        sql += ' WHERE studentnummer = ?';
+
+        DB.getasync(sql,studentID).then(((rows) => {if(rows.length == 0){
+            let insertSQL = "INSERT INTO Applications (studentnummer, voornaam, voorletters, tussenvoegsel, achternaam, geboortedatum, mobielnummer, studie)";
+            insertSQL+= " VALUES (?,?,?,?,?,?,?,?) ";
+            DB.insertAsync(insertSQL, [reqBody.fstudentnummer, reqBody.fvoornaam, reqBody.fvoorletters, reqBody.ftussenvoegsel, reqBody.fachternaam, reqBody.fgeboortedatum, reqBody.fmobiel, reqBody.fstudie])
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:105]: please double check if SQL injection is not possible

File: src/formBackend/RegistrationBackend.js [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+/**
+ * Async inserter for the database.
+ * @param sql The SQl for the query.
+ * @param parameters The parameters for the Query.
+ */
+DB.insertAsync = function (sql, parameters) {
+    return new Promise((resolve,reject) => {
+        DB.run(sql, parameters, function (error) {
+            if(error){
+                console.log("DATABASE ERROR: " + error);
```
Comment (Riscky) [inline src/formBackend/RegistrationBackend.js:81]: not sure if logging to console is going to be useful on a server, I hope this is written to the journal if run by systemctl

Comment (Mstiekema) [reply src/formBackend/RegistrationBackend.js:81]: `console.log`s will not be written anywhere except the console itself during runtime, so all logs will be lost once restarted

File: src/formBackend/RegistrationBackend.js (hunk 2 above)
Comment (THECASE99) [reply src/formBackend/RegistrationBackend.js:11]: It is an option in the form, therefore it is an option here.

File: src/formBackend/RegistrationBackend.js (hunk 11 above)
Comment (THECASE99) [reply src/formBackend/RegistrationBackend.js:105]: Double checked, SQL injections are not possible.

File: src/formBackend/RegistrationBackend.js (hunk 12 above)
Comment (THECASE99) [reply src/formBackend/RegistrationBackend.js:81]: It's just there for debugging now and to have a quick look when it is deployed. It is not inteneded to be stored.

File: src/formBackend/RegistrationBackend.js (hunk 4 above)
Comment (THECASE99) [reply src/formBackend/RegistrationBackend.js:22]: @KungFuDonkey can you check this with your frontend?

File: .gitignore (hunk 1 above)
Comment (THECASE99) [reply .gitignore:25]: Will do

**Pull request merged**